vlog4j-core/src/main/java/org/semanticweb/vlog4j/core/reasoner/implementation/VLogReasoner.java

```diff
@@ -808,4 +811,34 @@ ReasonerState getReasonerState() {
 	void setReasonerState(ReasonerState reasonerState) {
 		this.reasonerState = reasonerState;
 	}
+
+	@Override
+	public void getKbFacts(OutputStream stream) throws IOException {
+		// TODO Auto-generated method stub
+		HashSet<PositiveLiteral> headLiterals = new HashSet<PositiveLiteral>();
+		for (Rule rule : this.knowledgeBase.getRules()) {
+			for (PositiveLiteral positiveLiteral : rule.getHead()) {
+				headLiterals.add(positiveLiteral);
+			}
+		}
+		for (PositiveLiteral positiveliteral : headLiterals) {
+			try (final QueryResultIterator queryAnswers = this.answerQuery(positiveliteral, true)) {
+				while (queryAnswers.hasNext()) {
+					QueryResult queryAnswer = queryAnswers.next();
+					stream.write((positiveliteral.getPredicate().getName()
+							+ queryAnswer.getTerms().toString().replace("[", "(").replace("]", ").") + "\n")

```

Comment:
We should not have QueryResult String processing in Reasoner implementation.
Instead, this should be done in  QueryResult#toString() (which could delegate to Serializer).
Also, instead of replacing "[" with "(" in the String representation of a List, we should construct the String ourselves while iterating trough the list elements, and use the "(" and ")" constants from Serializer class.